### ISSUE TITLE
Table cell extraction not working as expected

### ISSUE THREAD

**Issue by theshahshow** [pymupdf/PyMuPDF#2875]
## Describe the bug (mandatory)

I am trying to extract table cells from a table, but when drawing the cells, more cells are actually getting detected which are not visible to human eye, by this i mean that you wouldn't guess a table cell there. According to my understanding, anything related to table extraction uses the pdf drawings (.get_drawings()). According to one github issue i saw, there may be some drawings which are not visible directly but are actually there. This is just my guess. If this is the case, then anyway to solve it? Or the problem is something else?

## To Reproduce (mandatory)
Explain the steps to reproduce the behavior, For example, include a minimal code snippet, example files, etc.

Code:
doc = fitz.open("AUTOSAR_SWS_CANNetworkManagement.pdf")
page = doc[97]
tables = check_table(page)

draw_table_cells(page, tables[2]) # this just draws the bounding boxes

Output:
<img width="566" alt="Screenshot 2023-12-09 at 7 53 53 AM" src="https://github.com/pymupdf/PyMuPDF/assets/68694517/ed0f71de-d303-4a42-be06-d1afcef38aa5">

One would think the "Multiplicity configuration class" would be 1 cell, but 3 cells are displayed there.

At first i thought because of the bold text, maybe how the algorithm works, it is merging the bottom parts of the bold text and hence we are getting bbox around those text. But then i tried to remove the text inside the table and then find the table cells. 

Code:

doc = fitz.open("AUTOSAR_SWS_CANNetworkManagement.pdf")
page = doc[97]
tables = check_table(page)

page.add_redact_annot(tables[2].bbox, "")
page.apply_redactions()

draw_table_cells(page, tables[2])

Output:
<img width="566" alt="Screenshot 2023-12-09 at 7 57 05 AM" src="https://github.com/pymupdf/PyMuPDF/assets/68694517/3c929d0d-b3b6-480f-b784-1f935b627d30">

The output is exactly the same, which means the problem was not with the text. Any ideas why is this happening?

**Comment by JorjMcKie** [comment]
Please provide the example file for issue reproduction.

**Comment by theshahshow** [comment]
[example_page.pdf](https://github.com/pymupdf/PyMuPDF/files/13646424/example_page.pdf)

In this page, tables 1, 2 and 3 show the same behavior as described before.

To figure out whats happening, i overlayed the output of page.get_drawings() on the empty tables. Here is the output:
[drawings_output.pdf](https://github.com/pymupdf/PyMuPDF/files/13646492/drawings_output.pdf)

It is infact the case that there are extra lines between the actual grid cells. For now, i am solving this problem by detecting vertical and horizontal lines from the empty table and forming cells using opencv and then mapping table cells detected by pymupdf to the cells detected by opencv. I also thought of redrawing the same table from page.get_drawings() thinking somehow i would uniquely identify those invisible lines, but was unable to do so.

Is this problem predictable or this problem can happen with any table and how the table was made or rendered by pdf? While solution using opencv works for now, i don't know if this is general or not.

**Comment by theshahshow** [comment]
Is there any way to get the line thickness of the drawings? The actual grid lines will be more thicker than these invisible lines.

**Comment by JorjMcKie** [comment]
1. As table detection is always a challenge, algorithms are grateful for every hint 😉.
2. Vector graphics extraction (`page.get_drawings()`), contains the line width ("width" key in spans of type "s" or "sf" are a float).
3. For a simple table (rows x columns) you can  specify the x- and / or y-coordinates of the lines (vertical, horizontal) you want to have considered. Then the detection will exclusively use those.

I am considering changes that would support a callback for delivering the line pieces ... with that there would be the option to read the vector graphics oneself or take whatever else into consideration.
I had users with white (invisible) rectangles existing underneath some parts of only some table cells. Because there is no way to detect a table as such and at the same time identify useless parts of the vector graphics, a callback routine may be helpful to filter out rectangles with certain fill colors ...

**Comment by JorjMcKie** [comment]
I am testing modifications / new options to the detection strategy.
Among them are a detection strategy that only accepts lines as table cell delimiters, and rectangles with stroked border lines. No rectangles with fill color only.

In your example page, if doing `find_tables(strategy="lines_strict")` then all rectangles having no stroked border are ignored.
Result:
```python
tabs = page.find_tables(strategy="lines_strict")
for i, tab in enumerate(tabs.tables):
    print(f"Table {i} {'-'*50}")
    for e in tab.extract():
        print(e)

        
Table 0 --------------------------------------------------
['Value Configuration Class', 'Pre-compile time', 'X', 'VARIANT-PRE-COMPILE']
['Link time', 'X', 'VARIANT-LINK-TIME, VARIANT-POST-\nBUILD']
['Post-build time', '--', '']
['Scope / Dependency', 'scope: local  \ndependency: (DRAFT) only valid if\nNmPnEraCalcEnabled == true \n \n(OBSOLETE) only valid if CanNmP\nCanNmPnEraCalcEnabled == true', 'NmPnEiraCalcEnabled == true or \nnEiraCalcEnabled == true or \n.']
Table 1 --------------------------------------------------
['SWS Item', 'ECUC_CanNm_00042 :']
['Name', 'CanNmBusLoadReductionActive']
['Parent Container', 'CanNmChannelConfig']
['Description', 'This parameter defines if bus load reduction for the respective NM channel \nis active or not.']
['Multiplicity', '1']
['Type', 'EcucBooleanParamDef']
['Default value', '--']
['Post-Build Variant Value', 'false']
['Value Configuration Class', 'Pre-compile time', 'X', 'VARIANT-PRE-COMPILE']
['Link time', 'X', 'VARIANT-LINK-TIME, VARIANT-POST-\nBUILD']
['Post-build time', '--', '']
['Scope / Dependency', 'scope: local  \ndependency: CanNmBusLoadReductionActive = false if \nCanNmBusLoadReductionEnabled == false']
Table 2 --------------------------------------------------
['SWS Item', 'ECUC_CanNm_00075 :']
['Name', 'CanNmCarWakeUpBitPosition']
['Parent Container', 'CanNmChannelConfig']
['Description', 'Specifies the Bit position of the CWU within the NM PDU.']
['Multiplicity', '0..1']
['Type', 'EcucIntegerParamDef']
['Range', '0 .. 7', '']
['Default value', '--']
['Post-Build Variant \nMultiplicity', 'false']
['Post-Build Variant Value', 'false']
['Multiplicity Configuration \nClass', 'Pre-compile time', 'X', 'VARIANT-PRE-COMPILE']
['Link time', 'X', 'VARIANT-LINK-TIME, VARIANT-POST-\nBUILD']
['Post-build time', '--', '']
['Value Configuration Class', 'Pre-compile time', 'X', 'VARIANT-PRE-COMPILE']
['Link time', 'X', 'VARIANT-LINK-TIME, VARIANT-POST-\nBUILD']
['Post-build time', '--', '']
['Scope / Dependency', 'scope: local  \ndependency: only available if Can', 'NmCarWakeUpRxEnabled == TRUE']
Table 3 --------------------------------------------------
['SWS Item', 'ECUC_CanNm_00076 :']
['Name', 'CanNmCarWakeUpBytePosition']
['Parent Container', 'CanNmChannelConfig']
['Description', 'Specifies the Byte position of the CWU within the NM PDU.']
['Multiplicity', '0..1']
['Type', 'EcucIntegerParamDef']
```
I think this indeed is better. The identified table cells are marked red in here. There still are two cases, where a visually undivided cell is sub-divided.
[x.pdf](https://github.com/pymupdf/PyMuPDF/files/13696156/x.pdf)

**Comment by theshahshow** [comment]
This makes sense. Why is the cell getting subdivided btw? There was no drawing originally there to begin with right?

Also, i tried this: Got the drawings using extended=True, filtered the drawings with type="clip" and then used those drawings to make the table again. From this new table i extracted the cells and the output was just fine. I have no idea why this works :|, but it worked for some reason.

**Comment by theshahshow** [comment]
Code:

doc = fitz.open("./Communication Services/AUTOSAR_SWS_CANNetworkManagement.pdf")
page = doc[91]

drawings = page.get_drawings(extended=True)[1:]

for drawing_elements in drawings:
    if drawing_elements["type"] == "clip":

        if drawing_elements["items"][0][0] == "re":
            rect_bbox = drawing_elements["items"][0][1]
            page.draw_rect(rect_bbox, color=(1, 0, 0))

Output: 
[modified_page.pdf](https://github.com/pymupdf/PyMuPDF/files/13696621/modified_page.pdf)

Red lines are the filtered rectangles.

**Comment by JorjMcKie** [comment]
This issue has been resolved with version 1.23.14.
